### PR TITLE
llext: MPU/MMU Support for Extensions

### PR DESCRIPTION
**Pull request by teburd**
~~Please ignore the first commit, its for my own needs as the zephyr python.cmake doesn't actually find the right python for me~~ see #65995

Supporting memory protection means allocating and assigning permissions
to memory partitions being allocated/referenced and providing those to a
memory domain. Unfortunately it seems adding partitions to the default
domain on arm cortex-m does not currently update the mpu regions, nor is
any of this possible without usermode enabled today. At the same time the test doesn't actually work to call hello world within a usermode thread as the extension struct itself... is not accessible. Ugh, will need a fix for that of course.

That said, this works and shows that the test can be run with user mode and MPU
enabled and gives a target to work on refining towards. I know @edersondisouza was looking for this and I believe @laurenmurphyx64 might need this soon. Certainly @lyakh will be needing it.

The idea currently is that each memory section (LLEXT_MEM) has a partition setup for it with the appropriate permissions needed. Really this can be simplified a bit though as we mostly need 3 partitions, an rx, ro, and rw partition. To clean that up I'll need to simplify the allocation setup to collect all the rx/ro/rw allocations needed and allocate, then place the sections appropriately.

Today this works though, and works even if you don't allocate and copy things like @lyakh's setup needs, at the cost of a per llext memory section k_mem_partition cost, or mpu region equivalent on arm cortex-m.

Runnable/buildable by doing something like the following...

```sh
west build -p always -b mimxrt1010_evk tests/subsys/llext -- -DCONFIG_USERSPACE=y
west flash -r pyocd
```

### CODE REVIEW

**Comment by laurenmurphyx64** [comment]
```
/home/lauren/zephyrproject/zephyr/tests/subsys/llext/src/test_llext_simple.c: In function 'llext_test_llext_simple':
/home/lauren/zephyrproject/zephyr/tests/subsys/llext/src/test_llext_simple.c:71:37: error: 'struct llext' has no member named 'mem_domain'
   71 |         k_mem_domain_add_thread(&ext->mem_domain, k_current_get());
      |                                     ^~
```
since mem_domain only exists for CONFIG_USERSPACE, so I think you need to fix up the #ifdef on test_llext_simple.c:70. [Here's some runs](https://gist.github.com/laurenmurphyx64/b7cf6d96a372961a03481d4cc978d882) of this PR on my FRDM, doesn't work (yet)

**Comment by teburd** [comment]
> ```
> /home/lauren/zephyrproject/zephyr/tests/subsys/llext/src/test_llext_simple.c: In function 'llext_test_llext_simple':
> /home/lauren/zephyrproject/zephyr/tests/subsys/llext/src/test_llext_simple.c:71:37: error: 'struct llext' has no member named 'mem_domain'
>    71 |         k_mem_domain_add_thread(&ext->mem_domain, k_current_get());
>       |                                     ^~
> ```
> 
> since mem_domain only exists for CONFIG_USERSPACE, so I think you need to fix up the #ifdef on test_llext_simple.c:70. [Here's some runs](https://gist.github.com/laurenmurphyx64/b7cf6d96a372961a03481d4cc978d882) of this PR on my FRDM, doesn't work (yet)

Ah it looks like the frdm k64f might not have enough mpu regions, at least the way this is written now. I bet if I reduce the partitions to the rx/ro/rw triple it'd work out. I'll get that done early next week.

The memory protection should probably be ifdef'ed on rather than userspace here frankly, but memory protection is tied to user space solely at the moment. LLEXT might be another thing that requires it.

Relevant log messages I saw there

```
E: no free partition slots available
E: failed to add partition 4 to domain
E: Failed to copy ELF sections, ret -28
```

**Review by pillo79 — COMMENTED** [review]
Looks good! My remarks on the current code:

* You should add logic to make sure that `CONFIG_LLEXT_STORAGE_WRITABLE` is not set, as 1) that skips the memory domain code, and 2) the buffers would not be correctly aligned anyway.

* I had [a branch that was working with memory domains](https://github.com/pillo79/zephyr/compare/4287d0b917266cf1b07b91ab818c5f7993b60cd0...dynamic_modules_support-hardened), but I was using separate static allocators for the RO, RW, and RX. The main thing that I was absolutely not able to support was `CONFIG_EXECUTE_XOR_WRITE` - there appeared to be no way to write to an area of memory that I wanted to use as executable (unless RWX was used). Is that config enabled in your MPU-allowed build? 

* `k_heap_aligned_alloc` will be leaving (possibly big) chunks of unused memory in the heap, for the few locations where alignment is possible. Is that sustainable with actual llext payload sizes? I think defining separate aligned allocators allows for more efficient use of the memory and also allows multiple llexts to be loaded.

File: include/zephyr/llext/llext.h [hunk 1]
```diff
@@ -30,8 +31,8 @@ extern "C" {
 enum llext_mem {
 	LLEXT_MEM_TEXT,
 	LLEXT_MEM_DATA,
-	LLEXT_MEM_RODATA,
 	LLEXT_MEM_BSS,
+	LLEXT_MEM_RODATA,
```
[the remark this answers is not in the document]
Comment (teburd) [reply include/zephyr/llext/llext.h:35]: I was thinking I could be smart about the way I iterate there for rx/rw/ro data segments, but after the weekend I've realized its probably better to iterate over the sections to gather all the memory needed for each then allocate once.

**Comment by teburd** [comment]
I've simplified this a bit, only 4 partitions are actually needed, and at best we might optimize to need 3 (two rw sections... .data and .bss) in the future.

Some work I think needs to happen to enable the memory protection APIs without CONFIG_USERSPACE=y

**Comment by teburd** [comment]
> Looks good! My remarks on the current code:
> 
>     * You should add logic to make sure that `CONFIG_LLEXT_STORAGE_WRITABLE` is not set, as 1) that skips the memory domain code, and 2) the buffers would not be correctly aligned anyway.

I don't think this config changes the need to setup memory partitions, particularly if you care about executing code from userspace. In particular for the RX .text section this is absolutely needed.

> 
>     * I had [a branch that was working with memory domains](https://github.com/pillo79/zephyr/compare/4287d0b917266cf1b07b91ab818c5f7993b60cd0...dynamic_modules_support-hardened), but I was using separate static allocators for the RO, RW, and RX. The main thing that I was absolutely not able to support was `CONFIG_EXECUTE_XOR_WRITE` - there appeared to be no way to write to an area of memory that I wanted to use as executable (unless RWX was used). Is that config enabled in your MPU-allowed build?

I tried this at first too but I think this route of setting up the partitions for what is allocated or already referenced (e.g. XIP type setups) is more reasonable, as allocation is not tied to the memory permissions then.

> 
>     * `k_heap_aligned_alloc` will be leaving (possibly big) chunks of unused memory in the heap, for the few locations where alignment is possible. Is that sustainable with actual llext payload sizes? I think defining separate aligned allocators allows for more efficient use of the memory and also allows multiple llexts to be loaded.

This does indeed leave large holes, as would using any mpu region. This could be optimized further for the allocation case by trying to do some intelligent packing of things where the power of 2 sizing (arm mpu) is needed. MMU just cares about aligning to pages and is simpler. That's assuming things are actually allocated and copied though, seemingly many use cases are referencing existing memory instead and don't require any allocation but will still need to setup the memory protection hardware appropriately.

At least that's my understanding of everything.

**Comment by pillo79** [comment]
Love how this nicely handles both MMU and MPU!
Small clarification re `CONFIG_LLEXT_STORAGE_WRITABLE`:
 
> I don't think this config changes the need to setup memory partitions, particularly if you care about executing code from userspace. In particular for the RX .text section this is absolutely needed.

I actually meant the opposite with my comment - currently, if you enable it, any section except BSS enters [lines 277-280](https://github.com/zephyrproject-rtos/zephyr/pull/67857/files#diff-af3b33d1fd9f53bfe92ba88f7a3941228401d26b9b9e44e80ff19f38133c17fdR277-R280) and will not set mem domains properly. To support this use case, ~buffers~ (nah, *ELF sections!*) need to be properly aligned either at link time or outside llext code, and that sounded to me overly complex to provide.

**Comment by teburd** [comment]
> Love how this nicely handles both MMU and MPU! Small clarification re `CONFIG_LLEXT_STORAGE_WRITABLE`:
> 
> > I don't think this config changes the need to setup memory partitions, particularly if you care about executing code from userspace. In particular for the RX .text section this is absolutely needed.
> 
> I actually meant the opposite with my comment - currently, if you enable it, any section except BSS enters [lines 277-280](https://github.com/zephyrproject-rtos/zephyr/pull/67857/files#diff-af3b33d1fd9f53bfe92ba88f7a3941228401d26b9b9e44e80ff19f38133c17fdR277-R280) and will not set mem domains properly. It's really complex to support this use case, since IIUC it would require ~buffers~ (nah, _ELF sections!_) to be properly aligned to mem domain boundaries either at link time or outside llext code. I'm not volounteering for that piece of CMake code! 😜

Indeed that's a @lyakh issue :-)

**Comment by teburd** [comment]
restructured the test to ensure that mpu/no mpu testing is done

No mmu tests are done as they don't work currently. Xtensa MMU has a strange fault @dcpleung @ceolin might be interested in trying this out.

Cortex-A9 isn't supported yet because the compilation of the extension isn't supported for Cortex-A of any kind yet.

**Comment by hakehuang** [comment]
> Some quirk with NXP's mpu on the kinetis k64f part doesn't let me add the partitions to the default domain (too many partitions already specified for this? how and why? maybe @dleach02 @danieldegrasse @hakehuang might know?)

NXP Kinetis series are using NXP modified MPU with below restrictions.

```
	/* The NXP MPU does not give precedence to memory regions like the ARM
	 * MPU, which means that if one region grants access then another
	 * region cannot revoke access. If an application enables hardware
	 * stack protection, we need to disable supervisor writes from the core
	 * to the stack guard region. As a result, we cannot have a single
	 * background region that enables supervisor read/write access from the
	 * core to the entire address space, and instead define two background
	 * regions that together cover the entire address space except for
	 * SRAM.
	 */
```

**Comment by teburd** [comment]
> > Some quirk with NXP's mpu on the kinetis k64f part doesn't let me add the partitions to the default domain (too many partitions already specified for this? how and why? maybe @dleach02 @danieldegrasse @hakehuang might know?)
> 
> NXP Kinetis series are using NXP modified MPU with below restrictions.
> 
> ```
> 	/* The NXP MPU does not give precedence to memory regions like the ARM
> 	 * MPU, which means that if one region grants access then another
> 	 * region cannot revoke access. If an application enables hardware
> 	 * stack protection, we need to disable supervisor writes from the core
> 	 * to the stack guard region. As a result, we cannot have a single
> 	 * background region that enables supervisor read/write access from the
> 	 * core to the entire address space, and instead define two background
> 	 * regions that together cover the entire address space except for
> 	 * SRAM.
> 	 */
> ```

That explains the larger number of static regions I believe that I was seeing on the frdm k64f and why I believe only 2 dynamic regions were available on the default domain, ouch.

So as the test will be a portable demonstration it creates a thread with a new domain and this works just fine. Something to maybe note somewhere though.

File: subsys/llext/llext.c [hunk 2]
```diff
@@ -770,6 +825,14 @@ static int do_llext_load(struct llext_loader *ldr, struct llext *ext,
 	ldr->sect_cnt = ldr->hdr.e_shnum;
 	ext->alloc_size += sect_map_sz;
 
+#ifdef CONFIG_USERSPACE
+	ret = k_mem_domain_init(&ext->mem_domain, 0, NULL);
+	if (ret != 0) {
```
Comment (ceolin) [inline subsys/llext/llext.c:830]: you can't initialize a memory domain without any partition AFAIK

File: subsys/llext/llext.c [hunk 3]
```diff
@@ -18,6 +17,13 @@ LOG_MODULE_REGISTER(llext, CONFIG_LLEXT_LOG_LEVEL);
 
 #include <string.h>
 
+#ifdef CONFIG_MMU_PAGE_SIZE
+#define LLEXT_PAGE_SIZE CONFIG_MMU_PAGE_SIZE
```
Comment (ceolin) [inline subsys/llext/llext.c:21]: This only matters if you have userspace enabled, otherwise this does not need to be page aligned right ? This will waste memory when not using userspace.

**Comment by ceolin** [comment]
> Update 3ish:
> 
> * Restructured the testcase.yaml to ensure that mpu/no mpu  testing and building is done
> * No mmu tests are done as they don't work currently.
>   
>   * qemu_xtensa_mmu has a strange fault @dcpleung @ceolin might be interested in trying this out.

I had to increase the heap to be able to allocate the needed memory, after that I got load prohibited exception that I need to look in more details. Will try to do this in the next days.

>   * qemu_cortex_a9 isn't supported yet because the compilation of the extension isn't supported for Cortex-A of any kind yet.
> * LLEXT_WRITABLE requires aligned sections and the test fails if I enable MPU (which would be expected) so no testing is done for WRITABLE && MPU/MMU enablement but the code is there to setup the partition assuming the memory is laid out correctly.

File: tests/subsys/llext/src/test_llext_simple.c [hunk 4]
```diff
@@ -45,7 +55,24 @@ ZTEST(llext, test_llext_simple)
...
+
+	k_thread_start(&llext_thread);
+
+	k_thread_join(&llext_thread, K_FOREVER);
+#else
+	k_mem_domain_add_thread(&ext->mem_domain, k_current_get());
+
+	llext_entry(ext, NULL, NULL);
+
+#endif
```
[the remark this answers is not in the document]
Comment (teburd) [reply tests/subsys/llext/src/test_llext_simple.c]: In theory its possible without the thread by adding the partitions to the default domain. This should work fine for MMU systems or MPU systems with enough regions available. In practice at a widespread testing level its not useful across all MPU systems as they don't all share the same number of dynamically available regions to program.

File: subsys/llext/llext.c (hunk 3 above)
Comment (teburd) [reply subsys/llext/llext.c:21]: That's not true as regions need to have memory attributes (r/w/x) assigned even without userspace around in this case, particularly the executable bit needs to be settable for the .text region such that its rx, and ideally the .rodata should be ro. Without being able to manipulate the page permissions the loadable module will fail to execute as the executable bit on the allocated or peeked at memory will not be setup correctly.

File: subsys/llext/llext.c (hunk 2 above)
Comment (teburd) [reply subsys/llext/llext.c:830]: the docs specifically say its allowed and note NULL may be used with 0 parts https://docs.zephyrproject.org/latest/kernel/usermode/memory_domain.html#c.k_mem_domain_init

**Review by pillo79 — CHANGES_REQUESTED** [review]
Looks nice overall, one tiny addition to make the `STORAGE_WRITABLE` case at least possible in theory (though we know it will currently fail in practice since buffers aren't aligned correctly). I had spotted this last week. but did not submit the review sooner :man_facepalming:

**Review by pillo79 — DISMISSED** [review]
@teburd I took the liberty of removing the Python commit as this was your intent in the PR description - please confirm. Otherwise LGTM, thanks for this!

File: subsys/llext/llext.c [hunk 5]
```diff
@@ -266,18 +304,37 @@ static int llext_copy_section(struct llext_loader *ldr, struct llext *ext,
...
 			return 0;
 		}
 	}
 
-	ext->mem[mem_idx] = k_heap_aligned_alloc(&llext_heap, sizeof(uintptr_t),
-						 ldr->sects[mem_idx].sh_size,
+#ifndef CONFIG_ARM_MPU
+	const uintptr_t sect_alloc = ROUND_UP(ldr->sects[mem_idx].sh_size, LLEXT_PAGE_SIZE);
+	const uintptr_t sect_align = LLEXT_PAGE_SIZE;
+#else
```
Comment (edersondisouza) [inline subsys/llext/llext.c:317]: Why is ARM MPU handled differently? Maybe some note so that anyone touching this in the future can keep it in mind?

File: subsys/llext/llext.c [hunk 6]
```diff
@@ -5,7 +5,6 @@
  *
  */
 
-#include "zephyr/sys/__assert.h"
```
Comment (edersondisouza) [inline subsys/llext/llext.c:8]: This seems unrelated - maybe it does belong to a cleanup patch?

File: tests/subsys/llext/hello_world/prj.conf [hunk 7]
```diff
@@ -1,7 +1,6 @@
 CONFIG_ZTEST=y
 CONFIG_ZTEST_STACK_SIZE=8192
 CONFIG_LOG=y
-CONFIG_LOG_MODE_IMMEDIATE=y
```
Comment (edersondisouza) [inline tests/subsys/llext/hello_world/prj.conf:4]: Unrelated?

File: tests/subsys/llext/hello_world/testcase.yaml [hunk 8]
```diff
@@ -3,20 +3,28 @@ common:
...
 tests:
   llext.simple.readonly:
     arch_exclude: xtensa # for now
+    filter: not CONFIG_MPU and not CONFIG_MMU and not CONFIG_SOC_SERIES_S32ZE_R52
     extra_configs:
       - arch:arm:CONFIG_ARM_MPU=n
       - CONFIG_LLEXT_STORAGE_WRITABLE=n
+  llext.simple.readonly_mpu:
+    arch_exclude: xtensa # for now
+    filter: CONFIG_CPU_HAS_MPU and not CONFIG_LOG_PRINTK
```
Comment (edersondisouza) [inline tests/subsys/llext/hello_world/testcase.yaml:17]: What is the issue with CONFIG_LOG_PRINTK? Maybe some note on commit message?

**Comment by teburd** [comment]
@edersondisouza great feedback, thanks for the review, will fix things up here

File: subsys/llext/llext.c (hunk 3 above)
Comment (ceolin) [reply subsys/llext/llext.c:21]: oh sorry, yep that is true. This needs to be page aligned.

File: subsys/llext/llext.c (hunk 5 above)
Comment (ceolin) [reply subsys/llext/llext.c:317]: Yep ... These alignments constraints are somehow already handled by userspace code, maybe we need to export this logic in future.

File: tests/subsys/llext/hello_world/testcase.yaml (hunk 8 above)
Comment (teburd) [reply tests/subsys/llext/hello_world/testcase.yaml:17]: LOG_PRINTK requires !USERSPACE and with LOG_PRINTK being forced on some builds the filter was needed to avoid CI errors

File: subsys/llext/llext.c (hunk 5 above)
Comment (teburd) [reply subsys/llext/llext.c:321]: I'd welcome a conveniently wrapped mmu page/mpu region allocator, I'd note the permission bits need to be set *after* copying of data occurs in this case.

**Review by edersondisouza — DISMISSED** [review]

**Comment by pillo79** [comment]
@teburd CI is failing on a dependency issue on `LOG_PRINTK`. Can you take a look?

```
warning: LOG_PRINTK (defined at subsys/logging/Kconfig.processing:8) has direct dependencies !USERSPACE && !LOG_MODE_MINIMAL && LOG with value n, but is currently being y-selected by the following symbols:
 - LOG_BACKEND_RTT_FORCE_PRINTK (defined at subsys/logging/backends/Kconfig.rtt:103), with value y, direct dependencies LOG_BACKEND_RTT && !LOG_FRONTEND_ONLY && !LOG_MODE_MINIMAL && LOG (value: y), and select condition LOG_BACKEND_RTT && !LOG_FRONTEND_ONLY && !LOG_MODE_MINIMAL && LOG (value: y)
 ```

**Review by pillo79 — APPROVED** [review]

**Review by edersondisouza — APPROVED** [review]
